models/Post.php

```diff
@@ -152,17 +152,19 @@ public function beforeSave()
      * Sets the "url" attribute with a URL to this object.
      * @param string $pageName
      * @param Controller $controller
+     * @param array $params Override request URL parameters
      *
      * @return string
      */
-    public function setUrl($pageName, $controller)
+    public function setUrl($pageName, $controller, $params = [])
     {
-        $params = [
-            'id'   => $this->id,
-            'slug' => $this->slug
-        ];
+        $category = $this->categories->count() ? $this->categories->first()->slug : null;
```

Comment:
@mjauvin Wrapping this with a check to see if `$params['category']` exists could potentially save us a couple of DB queries. No need to get the category if it's just gonna be overwritten.